File: src/main/java/nz/co/aetheric/sameness/Equatator.java
package nz.co.aetheric.sameness;

import javax.validation.constraints.NotNull;

/**
 * Defines a function that determines if two things are isEquatable.
 * <p>Author: <a href="http://gplus.to/tzrlk">Peter Cummuskey</a></p>
 * @param <T> The common class between both comparable objects.
 */
@SuppressWarnings("WeakerAccess")
public interface Equatator<T> {

	@NotNull
	Class<T> getType();

	<L extends T, R extends T> boolean isEquatable(L left, R right);

}

File: src/main/kotlin/nz/co/aetheric/sameness/Equatator.kt
package nz.co.aetheric.sameness

import javax.validation.constraints.NotNull

/**
 * Defines a function that determines if two things are isEquatable.
 *
 * Author: [Peter Cummuskey](http://gplus.to/tzrlk)
 * @param <T> The common class between both comparable objects.
 */
interface Equatator<T> {

	val type: Class<T>

	fun <L : T, R : T> isEquatable(left: L?, right: R?): Boolean

}
